In [1]:
import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

In [3]:
# one-hot code your target variable using the `to_categorical()` function from the `keras.utils` module
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

#### 1. In this task, you'll implement several ANN models with different activation functions. Specifically, do the following:

* Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the tanh activation function for each layer.

In [5]:
tanh_model = Sequential()
# our first dense layer
tanh_model.add(Dense(128, input_shape=(784,), activation="tanh"))
# our second dense layer
tanh_model.add(Dense(64, activation="tanh"))
# last layer is the output layer.
tanh_model.add(Dense(10, activation="softmax"))

# Compile the model

tanh_model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
tanh_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 15us/sample - loss: 1.0333 - accuracy: 0.7460
Epoch 2/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.5208 - accuracy: 0.8709s - loss: 0.5599 - ac
Epoch 3/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.4209 - accuracy: 0.8886
Epoch 4/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.3743 - accuracy: 0.8977
Epoch 5/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.3455 - accuracy: 0.9042
Epoch 6/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.3251 - accuracy: 0.9088
Epoch 7/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.3094 - accuracy: 0.9126
Epoch 8/20
60000/60000 [==============================] - 1s 10us/sample - loss: 0.2965 - accuracy: 0.9160
Epoch 9/20
60000/60000 [==============================] - 1s 10us/sample - loss: 0.2856 - accuracy: 

In [6]:
tanh_model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [7]:
# Print the model results
score = tanh_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.2088997276455164
Test accuracy: 0.9412


* Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using sigmoid activation function for each layer.

In [8]:
sig_model = Sequential()
# our first dense layer
sig_model.add(Dense(128, input_shape=(784,), activation="sigmoid"))
# our second dense layer
sig_model.add(Dense(64, activation="sigmoid"))
# last layer is the output layer.
sig_model.add(Dense(10, activation="softmax"))

sig_model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
sig_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 14us/sample - loss: 2.2943 - accuracy: 0.1646
Epoch 2/20
60000/60000 [==============================] - 1s 11us/sample - loss: 2.2352 - accuracy: 0.3303
Epoch 3/20
60000/60000 [==============================] - 1s 10us/sample - loss: 2.1753 - accuracy: 0.4966
Epoch 4/20
60000/60000 [==============================] - 1s 10us/sample - loss: 2.0908 - accuracy: 0.5769
Epoch 5/20
60000/60000 [==============================] - 1s 10us/sample - loss: 1.9678 - accuracy: 0.6114
Epoch 6/20
60000/60000 [==============================] - 1s 11us/sample - loss: 1.7995 - accuracy: 0.6347
Epoch 7/20
60000/60000 [==============================] - 1s 10us/sample - loss: 1.6010 - accuracy: 0.6593
Epoch 8/20
60000/60000 [==============================] - 1s 10us/sample - loss: 1.4066 - accuracy: 0.6844
Epoch 9/20
60000/60000 [==============================] - 1s 10us/sample - loss: 1.2421 - accuracy: 0.7081
Epoch 10/20
60

In [9]:
sig_model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_4 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_5 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [10]:
# Print the model results
score = sig_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.5682575039863587
Test accuracy: 0.8565


* Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the ReLU activation function for each layer.

In [11]:
# Create the model
relu_model = Sequential()

relu_model.add(Dense(128, input_shape=(784,), activation="relu"))
relu_model.add(Dense(64, activation="relu"))
relu_model.add(Dense(10, activation="softmax"))

In [12]:
relu_model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_7 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_8 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [13]:
# Compile the model
relu_model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# Fit the model
relu_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 14us/sample - loss: 1.2247 - accuracy: 0.6756
Epoch 2/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.4986 - accuracy: 0.8680
Epoch 3/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.3872 - accuracy: 0.8919
Epoch 4/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.3402 - accuracy: 0.9036
Epoch 5/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.3120 - accuracy: 0.9115
Epoch 6/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.2913 - accuracy: 0.9168
Epoch 7/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.2750 - accuracy: 0.9214
Epoch 8/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.2611 - accuracy: 0.9252
Epoch 9/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.2492 - accuracy: 0.9287
Epoch 10/20
60

In [14]:
# Print the model results
score = relu_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.16815965002402664
Test accuracy: 0.951


* tanh = 94%
* sigmoid = 86%
* reLU = 95%

The reLU model performed the best

#### 2. In this task, you'll implement the ANN models specified below. For each, use the hinge loss function as the loss function.

* Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the tanh activation function for each layer.

In [15]:
# Create the model
tanh_model = Sequential()

tanh_model.add(Dense(128, input_shape=(784,), activation="tanh"))
tanh_model.add(Dense(64, activation="tanh"))
tanh_model.add(Dense(10, activation="softmax"))

# Compile the model
tanh_model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

# Fit the model
tanh_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.9905 - accuracy: 0.4075
Epoch 2/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.7830 - accuracy: 0.6694
Epoch 3/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.6031 - accuracy: 0.7951
Epoch 4/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.4744 - accuracy: 0.8434
Epoch 5/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.3946 - accuracy: 0.8637
Epoch 6/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.3457 - accuracy: 0.8748
Epoch 7/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.3133 - accuracy: 0.8827
Epoch 8/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.2901 - accuracy: 0.8887
Epoch 9/20
60000/60000 [==============================] - 1s 13us/sample - loss: 0.2725 - accuracy: 0.8938
Epoch 10/20
60

In [16]:
# Print the model results
score = tanh_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.18412239052653312
Test accuracy: 0.9228


* Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the sigmoid activation function for each layer.

In [17]:
# Create the model
sig_model = Sequential()

sig_model.add(Dense(128, input_shape=(784,), activation="sigmoid"))
sig_model.add(Dense(64, activation="sigmoid"))
sig_model.add(Dense(10, activation="softmax"))

# Compile the model
sig_model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

# Fit the model
sig_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 15us/sample - loss: 1.0183 - accuracy: 0.1598
Epoch 2/20
60000/60000 [==============================] - 1s 12us/sample - loss: 1.0029 - accuracy: 0.1968
Epoch 3/20
60000/60000 [==============================] - 1s 13us/sample - loss: 1.0022 - accuracy: 0.2487
Epoch 4/20
60000/60000 [==============================] - 1s 13us/sample - loss: 1.0016 - accuracy: 0.3076
Epoch 5/20
60000/60000 [==============================] - 1s 12us/sample - loss: 1.0011 - accuracy: 0.3643
Epoch 6/20
60000/60000 [==============================] - 1s 12us/sample - loss: 1.0006 - accuracy: 0.4105
Epoch 7/20
60000/60000 [==============================] - 1s 13us/sample - loss: 1.0002 - accuracy: 0.4462
Epoch 8/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.9998 - accuracy: 0.4756
Epoch 9/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.9994 - accuracy: 0.5007
Epoch 10/20
60

In [18]:
# Print the model results
score = sig_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.9955600111007691
Test accuracy: 0.6541


* Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the ReLU activation function for each layer.

In [19]:
# Create the model
relu_model = Sequential()

relu_model.add(Dense(128, input_shape=(784,), activation="relu"))
relu_model.add(Dense(64, activation="relu"))
relu_model.add(Dense(10, activation="softmax"))

# Compile the model
relu_model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

# Fit the model
relu_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 15us/sample - loss: 1.0128 - accuracy: 0.2975
Epoch 2/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.9793 - accuracy: 0.5103
Epoch 3/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.8938 - accuracy: 0.5912
Epoch 4/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.7352 - accuracy: 0.6821
Epoch 5/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.5554 - accuracy: 0.7706
Epoch 6/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.4334 - accuracy: 0.8490
Epoch 7/20
60000/60000 [==============================] - 1s 12us/sample - loss: 0.3501 - accuracy: 0.8763
Epoch 8/20
60000/60000 [==============================] - 1s 13us/sample - loss: 0.3005 - accuracy: 0.8866
Epoch 9/20
60000/60000 [==============================] - 1s 11us/sample - loss: 0.2699 - accuracy: 0.8940
Epoch 10/20
60

In [20]:
# Print the model results
score = relu_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.1635830114901066
Test accuracy: 0.9291


* Compare the results of each model with the result of the same model from the previous task. Which loss function performed best?

With crossentropy loss:
* tanh = 94%
* sigmoid = 86%
* reLU = 95%

With hinge loss:
* tanh = 92%
* sigmoid = 66%
* reLU = 92%

* All 3 activation types performed worse using hinge loss
* tanh and reLU performed the best using either crossentropy or hinge loss